docs/tutorials/course/fairbo.md

```diff
@@ -0,0 +1,506 @@
+    result = result.assign(fair_best=result["fair_error"].cummin())
+    return result
+```
+
+
+```python
+ACTIVE_METRIC_NAME = 'obj_metric'
+CONSTRAINT_METRIC_NAME = 'constr_metric'
+REWARD_ATTR_NAME = 'objective'
+FAIRNESS_THRESHOLD = 0.01
+FAIRNESS_DEFINITION = DSP  # You can use any fairness definition, such as DSP or DEO
+```
+
+We tune RF on a 4-dimensional search space: 
+
+* number of trees in {1, 2, . . . , 64}
+* min_samples_split in [0.01, 0.5] (log scaled)
+* tree maximum depth in {1, 2, 3, 4, 5}
+* criterion for quality of split in {Gini, Entropy}
+
+
+```python
+def create_train_fn_constraint(fairness_threshold, fairness_definition):
+    @ag.args(n_estimators=ag.space.Int(lower=1, upper=128),
+             min_samples_split=ag.space.Real(lower=0.01, upper=1.0, log=True),
+             max_depth=ag.space.Int(lower=1, upper=50),
+             criterion=ag.space.Categorical('gini', 'entropy')
+            )
+    def run_opaque_box(args, reporter):
+        opaque_box_eval = opaque_box(args.n_estimators, 
+                                     args.min_samples_split,
+                                     args.max_depth, 
+                                     args.criterion, 
+                                     fairness_threshold,
+                                     fairness_definition)
+        reporter(objective=-opaque_box_eval[ACTIVE_METRIC_NAME],
+                 constraint_metric=opaque_box_eval[CONSTRAINT_METRIC_NAME])
+    return run_opaque_box
+
+run_opaque_box = create_train_fn_constraint(fairness_threshold=FAIRNESS_THRESHOLD, 
+                                            fairness_definition=FAIRNESS_DEFINITION)  
+# fairness constraint: DSP < 0.01
+```
+
+
+```python
+def opaque_box(n_estimators, min_samples_split, max_depth, criterion, fairness_threshold, fairness_definition):
+    classifier = RandomForestClassifier(n_estimators=n_estimators, 
+                                        max_depth=max_depth,
+                                        min_samples_split=min_samples_split,
+                                        criterion=criterion)
+    classifier.fit(X_train, Y_train)
+    DSP_ForeignWorker = fairness_definition(classifier, X_test, Y_test, df["ForeignWorker"].values == "yes")
+    accuracy = accuracy_score(classifier.predict(X_test), Y_test)
+    evaluation_dict = {}
+    evaluation_dict[ACTIVE_METRIC_NAME] = accuracy
+    evaluation_dict[CONSTRAINT_METRIC_NAME] = DSP_ForeignWorker - fairness_threshold  # If DSP < fairness threshold, then fair
+    return evaluation_dict
+```
+
+
+```python
+# Create scheduler and searcher:
+
+# First get_config are random, the remaining ones use constrained BO
+search_options = {
+    'random_seed': SEED,
+    'num_fantasy_samples': 5,
+    'num_init_random': 1,
+    'debug_log': True}
+myscheduler = ag.scheduler.FIFOScheduler(
+    run_opaque_box,
+    searcher='bayesopt',
+    search_options=search_options,
+    num_trials=15,
```

Comment:
As in our other tutorials, I'd set this to a small value like 3-5, and add comment that "this only done for quick demo, and you should try  larger values"